### ISSUE TITLE
D3D12: Dual source blending behaves wrongly if the second source is only used on a disabled output

### ISSUE THREAD

**Issue by Pokechu22** [IGCIT/Intel-GPU-Community-Issue-Tracker-IGCIT#214]
### Checklist [README]

- [X] Device is not undervolted nor overclocked
- [X] Device is using the [latest drivers](https://github.com/IGCIT/Intel-GPU-Community-Issue-Tracker-IGCIT/wiki/Intel-driver-updates-and-changelogs)
- [X] Application is not cracked, modded and use the latest patch

### Application [Required]

Dolphin Emulator

### Processor / Processor Number [Required]

Intel Core i7-10750H CPU

### GPU Driver Version [Required]

- [ ] 31.0.101.2115 (*latest*)
- [X] 31.0.101.2114

### Rendering API [Required]

- [ ] Vulkan
- [ ] OpenGL
- [X] DirectX12
- [ ] DirectX11
- [ ] DirectX10
- [ ] DirectX9
- [ ] Not applicable

### Windows Build Number [Required]

- [ ] Windows 11 22H2
- [ ] Windows 11 21H2
- [ ] Windows 10 22H2
- [X] Windows 10 21H2
- [ ] Other (Please specify)

### Intel System Support Utility report

[igcit_ssu.txt](https://github.com/IGCIT/Intel-GPU-Community-Issue-Tracker-IGCIT/files/10349361/igcit_ssu.txt)

As a side note the anonymization doesn't seem to remove `Network Name`/`Profile`, nor does it remove it remove the username/machine name from log messages (e.g. "A new BITS job could not be created"). I haven't manually removed them either as I don't think it's an issue in my case, but this is something you may want to change.

The "The system has rebooted without cleanly shutting down first." line at the top was due to me needing to force shut down my laptop after updating the driver (I was previously on 30.0.100.9684); after the updating finished I was left with a completely black screen. I'm not entirely sure why that happened, but things were fine after it booted again.

### Description and steps to reproduce [Required]

Super Mario Sunshine draws its skyboxes by first drawing clouds and sun flare in greyscale and then drawing the sky afterwards with a source factor of 1 and a dest factor of 1-src_color (this results in the clouds being tinted blue). This process is done on an RGB888 framebuffer, so alpha update is disabled. Dolphin converts this to `SrcBlend = D3D12_BLEND_ONE, DestBlend = D3D12_BLEND_INV_SRC_COLOR, SrcBlendAlpha = D3D12_BLEND_ONE, DestBlendAlpha = D3D12_BLEND_INV_SRC1_ALPHA, RenderTargetWriteMask = D3D12_COLOR_WRITE_ENABLE_RED | D3D12_COLOR_WRITE_ENABLE_GREEN | D3D12_COLOR_WRITE_ENABLE_BLUE`. In this case, the Intel driver does the wrong thing, acting as if the source factor were 0 and the dest factor 1. This *only* affects D3D12, and also does not affect my Nvidia GPU.

I have found a few things that give correct results:
* Setting `DestBlendAlpha` to `D3D12_BLEND_ONE` instead of `D3D12_BLEND_INV_SRC1_ALPHA`
* Setting `SrcBlend` and `SrcBlendAlpha` to `D3D12_BLEND_SRC1_ALPHA` (in this case, that is always 1, although that is not generally true)
  * `D3D12_BLEND_INV_DEST_ALPHA` does not work (even though dest alpha is always zero in this case)
* Enabling `D3D12_COLOR_WRITE_ENABLE_ALPHA` (which is also fine here, but is not a general solution).

The `DestBlendAlpha` approach seems like it should be free of side effects, and is what I'm doing in https://github.com/dolphin-emu/dolphin/pull/11406. However, this still seems like a driver bug that behaviors differ depending on `SrcBlendAlpha`/`DestBlendAlpha` when `D3D12_COLOR_WRITE_ENABLE_ALPHA` is not enabled.

Here are detailed reproduction steps:

* Download Dolphin (e.g. the latest beta build, [5.0-17995](https://dolphin-emu.org/download/dev/8bad821019721b9b72701b495da95656ace5fea5/))
* Download and unzip [SMSSkyFifologs.zip](https://github.com/dolphin-emu/dolphin/files/10349115/SMSSkyFifologs.zip) which are captures of the relevant GameCube graphics commands which can be replayed in dolphin
* Launch dolphin; select config and uncheck "enable dual core"
* Close the config window, then select graphics
* Set Backend to Direct3D 12, and adapter to Intel(R) UHD Graphics
* Close the graphics window
* Select Tools -> Fifo Player
* Select load, then select SMSSky.dff
* Observe the black sky

There are screenshots on https://github.com/dolphin-emu/dolphin/pull/11406.

**Comment by IntelSupport-Rozilah** [comment]
hi @Pokechu22,
Thanks for the clear explanation, we have started to verify the issue and couldn't find the screenshot from the link you provided.
please confirm that image I attached is like the issue you observed.
1. black sky issue observed(tested with Intel UHD graphic)
![image](https://user-images.githubusercontent.com/44628772/210951314-598d2c74-9c54-457b-9961-e6bb51ddd18c.png)
2. issue is not observed and a blue sky appears correctly.(tested with Nvidia RTX 3060)
![image](https://user-images.githubusercontent.com/44628772/210951856-06428c4a-db02-45ca-911d-692c4c0b807e.png)

**Comment by Pokechu22** [comment]
Yes, those images match what I am seeing. On my link, the images are inside of [`<details>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/details) tags which you need to click on to show the screenshots though; sorry for the lack of clarity about that.

**Comment by IntelSupport-Rozilah** [comment]
hi @Pokechu22,
The case has been verified on our side and we are able to reproduce it with our configurations as below:
Processor: Intel(R) Core(TM) i7-11700K
OS build: Windows 10 Home 19044(21H2)
Graphic device: Intel UHD Graphic 750
Graphic Driver version: 31.0.101.3790
We will be working on a fix. I want to give you a heads-up that the fix may take 3 to 6 months to be included.

**Comment by Pokechu22** [comment]
Thanks for the heads-up. I'll look into getting the workaround of changing `SrcBlendAlpha`/`DestBlendAlpha` to Zero/One if updates are disabled (https://github.com/dolphin-emu/dolphin/pull/11406) merged in the mean time. If we discover any situations where that workaround doesn't work I'll let you know. I'll edit the comment when the workaround is merged. (This does also mean that when confirming it's fixed, you'll want to use a version without the workaround, such as [5.0-17995](https://dolphin-emu.org/download/dev/8bad821019721b9b72701b495da95656ace5fea5/)).

**Comment by SavageWaffle** [comment]
@Felipe-Intel Is there any update on this? Looks like it was merged into a ticket with Dolphin Emulator.

**Comment by Pokechu22** [comment]
This is still present with driver version 31.0.101.2134 and Dolphin 5.0-18498, but it doesn't affect newer versions of Dolphin due to the workaround mentioned above.

**Comment by SavageWaffle** [comment]
@Pokechu22 It sounds like this was fixed on Dolphins side and can be resolved? Or is there a need for it to work on prior versions of Dophin?

**Comment by SavageWaffle** [comment]
@Felipe-Intel I think this is solved

**Comment by Felipe-Intel** [comment]
@Pokechu22 I just noticed this was originally reported on a 10th gen processor. 10th gen and older platforms transitioned to a legacy support model a while ago. You can find more details [in this article](https://www.intel.com/content/www/us/en/support/articles/000091662.html). I'll proceed to close this ticket. 
If you find that the issue is still happening with the latest version of the software in our newer platforms (e.g. Arc Graphics) feel free to open a new ticket. Considering this is an emulator and to avoid copyright violations, please provide a sample code where the issue is reproducible. 
cc @mujeeb-intel
